llama_index/indices/query/query_transform/feedback_transform.py

```diff
@@ -1,37 +1,105 @@
+import logging
 from typing import Dict, Optional
 
 from llama_index.evaluation.base import Evaluation
 from llama_index.indices.query.query_transform.base import BaseQueryTransform
 from llama_index.indices.query.schema import QueryBundle
+from llama_index.langchain_helpers.chain_wrapper import LLMPredictor
+from llama_index.llm_predictor.base import BaseLLMPredictor
+from llama_index.prompts.base import Prompt
+from llama_index.response.schema import Response
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RESYNTHESIS_PROMPT_TMPL = (
+    "Here is the original query:\n"
+    "{query_str}\n"
+    "Here is the response given:\n"
+    "{response}\n"
+    "Here is some feedback from evaluator about the response given.\n"
+    "{feedback}\n"
+    "If you want to resynthesize the query, please return the modified query below.\n"
+    "Otherwise, please return the original query.\n"
+)
+
+DEFAULT_RESYNTHESIS_PROMPT = Prompt(DEFAULT_RESYNTHESIS_PROMPT_TMPL)
 
 
 class FeedbackQueryTransformation(BaseQueryTransform):
     """Transform the query given the evaluation feedback.
 
     Args:
         eval(Evaluation): An evaluation object.
+        llm_predictor(BaseLLMPredictor): An LLM predictor.
+        resynthesize_query(bool): Whether to resynthesize the query.
+        resynthesis_prompt(Prompt): A prompt for resynthesizing the query.
 
     """
 
     def __init__(
         self,
-        evaluation: Evaluation,
+        evaluation: Optional[Evaluation] = None,
+        llm_predictor: Optional[BaseLLMPredictor] = None,
+        resynthesize_query: bool = False,
+        resynthesis_prompt: Optional[Prompt] = None,
     ) -> None:
         super().__init__()
         self.evaluation = evaluation
+        self.llm_predictor = llm_predictor or LLMPredictor()
+        self.should_resynthesize_query = resynthesize_query
+        self.resynthesis_prompt = resynthesis_prompt or DEFAULT_RESYNTHESIS_PROMPT
+
+    def set_eval(self, evaluation: Evaluation) -> None:
+        """Set evaluation."""
+        self.evaluation = evaluation
 
     def _run(self, query_bundle: QueryBundle, extra_info: Dict) -> QueryBundle:
-        query_str = query_bundle.query_str
-        new_query = (
-            query_str
-            + "\n----------------\n"
-            + self.construct_feedback(self.evaluation.response.response)
-        )
-        return QueryBundle(new_query)
-
-    def construct_feedback(self, response: Optional[str]) -> str:
+        orig_query_str = query_bundle.query_str
+        if self.evaluation is None:
+            raise ValueError("Evaluation is not set.")
+        if self.evaluation.feedback == "YES" or self.evaluation.feedback == "NO":
```

Comment:
very nit: why don't we use passing flag instead of this str check?

Reply:
Oh nvm I think I get it now. This is because different evaluator could give different responses.

Reply:
hmm, this seems like an abuse of the `feedback` field to just say yes or no.

Reply:
This is to maintain backward compatibility with the behavior of QueryResponseEvaluator. The old behavior only returns "YES" or "NO", for newer versions we have actual feedback. I keep the check for the old feedback because we want different behavior.